File: python/taco/reader/versioned.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from typing import Any, cast
from urllib.error import HTTPError, URLError
from urllib.parse import urljoin, urlsplit
from urllib.request import urlopen

from .._source import Location, Source, normalize
from ..contract.collection import SEMVER
from ..errors import ContainerError


@dataclass(frozen=True)
class Resolution:
    sources: tuple[Location, ...]
    collection: dict[str, Any] | None = None
    version: str | None = None
    versions: tuple[str, ...] = ()
    manifest: Location | None = None


def _direct(name: str) -> bool:
    return name == ".tacocat" or name.lower().endswith(".zip") or SEMVER.fullmatch(name) is not None


def _candidate(source: Location) -> Location | None:
    if isinstance(source, Path):
        if source.name == "taco.json":
            return source
        if _direct(source.name):
            return None
        candidate = source / "taco.json"
        return candidate if candidate.is_file() else None

    parsed = urlsplit(source)
    name = PurePosixPath(parsed.path).name
    if name == "taco.json":
        return source
    if _direct(name):
        return None
    return urljoin(source.rstrip("/") + "/", "taco.json")


def _read(candidate: Location, *, required: bool) -> bytes | None:
    if isinstance(candidate, Path):
        try:
            return candidate.read_bytes()
        except FileNotFoundError as error:
            if not required:
                return None
            raise ContainerError(f"versioned manifest does not exist: {candidate}") from error
        except OSError as error:
            raise ContainerError(f"could not read versioned manifest {candidate}: {error}") from error

    try:
        with urlopen(candidate) as response:
            return cast(bytes, response.read())
    except HTTPError as error:
        if error.code == 404 and not required:
            return None
        raise ContainerError(f"could not read versioned manifest {candidate}: HTTP {error.code}") from error
    except URLError as error:
        raise ContainerError(f"could not read versioned manifest {candidate}: {error.reason}") from error


def _object(value: object, *, context: str) -> dict[str, Any]:
    if not isinstance(value, dict) or not all(isinstance(key, str) for key in value):
        raise ContainerError(f"{context} must be an object")
    return value


def _parse(payload: bytes, candidate: Location) -> dict[str, Any]:
    try:
        value = json.loads(payload)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ContainerError(f"versioned manifest is not valid JSON: {candidate}") from error
    manifest = _object(value, context="versioned manifest")
    if manifest.get("taco:container") != "versioned":
        raise ContainerError("taco:container must be 'versioned'")
    return manifest


def _join(candidate: Location, href: str) -> Location:
    if "://" in href:
        return href
    if isinstance(candidate, Path):
        return (candidate.parent / href).resolve()
    return urljoin(candidate, href)


def resolve(source: Source) -> Resolution:
    paths = normalize(source)
    if len(paths) != 1:
        return Resolution(paths)

    original = paths[0]
    explicit_manifest = (isinstance(original, Path) and original.name == "taco.json") or (
        isinstance(original, str) and PurePosixPath(urlsplit(original).path).name == "taco.json"
    )
    candidate = _candidate(original)
    if candidate is None:
        return Resolution(paths)

    payload = _read(candidate, required=explicit_manifest)
    if payload is None:
        return Resolution(paths)
    manifest = _parse(payload, candidate)
    versions = _object(manifest.get("taco:versions"), context="taco:versions")
    if not versions:
        raise ContainerError("taco:versions must contain at least one version")

    selected = manifest.get("taco:default_version")
    if not isinstance(selected, str) or not selected:
        raise ContainerError("taco:default_version must be a non-empty string")
    if selected not in versions:
        raise ContainerError(f"taco:default_version {selected!r} is not present in taco:versions")

    entries: dict[str, tuple[str, dict[str, Any]]] = {}
    for version, value in versions.items():
        if SEMVER.fullmatch(version) is None:
            raise ContainerError(f"taco:versions key {version!r} must follow Semantic Versioning")
        entry = _object(value, context=f"version {version!r}")
        href = entry.get("href")
        if not isinstance(href, str) or not href:
            raise ContainerError(f"version {version!r} needs a non-empty href")
        collection = _object(entry.get("collection"), context=f"version {version!r} collection")
        dataset_version = collection.get("dataset_version")
        if dataset_version != version:
            raise ContainerError(f"version {version!r} embeds collection dataset_version {dataset_version!r}")
        entries[version] = (href, collection)

    href, collection = entries[selected]

    return Resolution(
        (_join(candidate, href),),
        collection=collection,
        version=selected,
        versions=tuple(versions),
        manifest=candidate,
    )


__all__ = ["Resolution", "resolve"]

File: r/R/versioned.R
.SEMVER <- paste0(
  "^(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)",
  "(?:-[0-9A-Za-z-]+(?:\\.[0-9A-Za-z-]+)*)?",
  "(?:\\+[0-9A-Za-z-]+(?:\\.[0-9A-Za-z-]+)*)?$"
)


.source_name <- function(source) {
  path <- sub("[?#].*$", "", sub("/+$", "", source))
  basename(path)
}


.direct_source <- function(source) {
  name <- .source_name(source)
  identical(name, ".tacocat") ||
    grepl("\\.zip$", name, ignore.case = TRUE) ||
    grepl(.SEMVER, name, perl = TRUE)
}


.is_url <- function(source) {
  grepl("^[A-Za-z][A-Za-z0-9+.-]*://", source)
}


.manifest_candidate <- function(source) {
  if (identical(.source_name(source), "taco.json")) {
    return(source)
  }
  if (.direct_source(source)) {
    return(NULL)
  }
  if (.is_url(source)) {
    clean <- sub("[?#].*$", "", source)
    return(paste0(sub("/+$", "", clean), "/taco.json"))
  }
  candidate <- file.path(source, "taco.json")
  if (file.exists(candidate)) candidate else NULL
}


.read_manifest <- function(candidate, required) {
  if (!.is_url(candidate)) {
    if (!file.exists(candidate)) {
      if (!required) return(NULL)
      .taco_stop("versioned manifest does not exist: %s", candidate)
    }
    return(tryCatch(
      readBin(candidate, what = "raw", n = file.info(candidate)$size),
      error = function(err) {
        .taco_stop("could not read versioned manifest %s: %s", candidate, conditionMessage(err))
      }
    ))
  }

  response <- tryCatch(
    curl::curl_fetch_memory(
      candidate,
      handle = curl::new_handle(followlocation = TRUE)
    ),
    error = function(err) {
      .taco_stop("could not read versioned manifest %s: %s", candidate, conditionMessage(err))
    }
  )
  if (identical(response$status_code, 404L) && !required) {
    return(NULL)
  }
  if (response$status_code < 200L || response$status_code >= 300L) {
    .taco_stop(
      "could not read versioned manifest %s: HTTP %d",
      candidate,
      response$status_code
    )
  }
  response$content
}


.json_object <- function(value, context) {
  if (!is.list(value) || (length(value) && (is.null(names(value)) || any(!nzchar(names(value)))))) {
    .taco_stop("%s must be an object", context)
  }
  value
}


.parse_manifest <- function(payload, candidate) {
  manifest <- tryCatch(
    jsonlite::fromJSON(rawToChar(payload), simplifyVector = FALSE),
    error = function(err) {
      .taco_stop("versioned manifest is not valid JSON: %s", candidate)
    }
  )
  manifest <- .json_object(manifest, "versioned manifest")
  if (!identical(manifest[["taco:container"]], "versioned")) {
    .taco_stop("taco:container must be 'versioned'")
  }
  manifest
}


.normalize_url_path <- function(path) {
  trailing <- endsWith(path, "/")
  result <- character()
  for (segment in strsplit(path, "/", fixed = TRUE)[[1L]]) {
    if (!nzchar(segment) || identical(segment, ".")) next
    if (identical(segment, "..")) {
      if (length(result)) result <- result[-length(result)]
    } else {
      result <- c(result, segment)
    }
  }
  normalized <- paste0("/", paste(result, collapse = "/"))
  if (trailing && normalized != "/") normalized <- paste0(normalized, "/")
  normalized
}


.join_manifest_href <- function(candidate, href) {
  if (.is_url(href)) {
    return(href)
  }
  if (!.is_url(candidate)) {
    if (grepl("^[/\\\\]", href)) {
      return(normalizePath(sub("[/\\\\]+$", "", href), mustWork = FALSE))
    }
    joined <- file.path(dirname(candidate), href)
    return(normalizePath(sub("[/\\\\]+$", "", joined), mustWork = FALSE))
  }

  clean <- sub("[?#].*$", "", candidate)
  if (startsWith(href, "//")) {
    scheme <- sub("^([A-Za-z][A-Za-z0-9+.-]*):.*$", "\\1", clean)
    return(paste0(scheme, ":", href))
  }
  if (startsWith(href, "/")) {
    origin <- sub("^([A-Za-z][A-Za-z0-9+.-]*://[^/]+).*$", "\\1", clean)
    return(paste0(origin, .normalize_url_path(href)))
  }
  origin <- sub("^([A-Za-z][A-Za-z0-9+.-]*://[^/]+).*$", "\\1", clean)
  candidate_path <- sub("^[A-Za-z][A-Za-z0-9+.-]*://[^/]+", "", clean)
  if (!nzchar(candidate_path)) candidate_path <- "/"
  target <- paste0(sub("[^/]*$", "", candidate_path), href)
  paste0(origin, .normalize_url_path(target))
}


.validate_embedded_collection <- function(collection, version) {
  required <- c(
    "taco:version", "id", "dataset_version", "description", "licenses",
    "providers", "tasks", "taco:structure", "taco:metadata"
  )
  missing <- setdiff(required, names(collection))
  if (length(missing)) {
    .taco_stop(
      "version '%s' embeds an invalid collection: missing required fields %s",
      version,
      paste(missing, collapse = ", ")
    )
  }
  if (!identical(collection[["taco:version"]], "3.0.0")) {
    .taco_stop("version '%s' embeds an invalid collection: taco:version must be 3.0.0", version)
  }
  if (!is.list(collection[["taco:metadata"]])) {
    .taco_stop("version '%s' embeds an invalid collection: taco:metadata must be an object", version)
  }
  invisible(collection)
}


.resolve_versioned <- function(source) {
  sources <- .check_source(source)
  direct <- list(
    sources = sources,
    collection = NULL,
    version = NULL,
    versions = character(),
    manifest = NULL
  )
  if (length(sources) != 1L) {
    return(direct)
  }

  original <- sources[[1L]]
  explicit <- identical(.source_name(original), "taco.json")
  candidate <- .manifest_candidate(original)
  if (is.null(candidate)) {
    return(direct)
  }
  payload <- .read_manifest(candidate, required = explicit)
  if (is.null(payload)) {
    return(direct)
  }

  manifest <- .parse_manifest(payload, candidate)
  versions <- .json_object(manifest[["taco:versions"]], "taco:versions")
  if (!length(versions)) {
    .taco_stop("taco:versions must contain at least one version")
  }
  selected <- manifest[["taco:default_version"]]
  if (!is.character(selected) || length(selected) != 1L || !nzchar(selected)) {
    .taco_stop("taco:default_version must be a non-empty string")
  }
  if (!selected %in% names(versions)) {
    .taco_stop("taco:default_version '%s' is not present in taco:versions", selected)
  }

  entries <- vector("list", length(versions))
  names(entries) <- names(versions)
  for (version in names(versions)) {
    if (!grepl(.SEMVER, version, perl = TRUE)) {
      .taco_stop("taco:versions key '%s' must follow Semantic Versioning", version)
    }
    entry <- .json_object(versions[[version]], sprintf("version '%s'", version))
    href <- entry[["href"]]
    if (!is.character(href) || length(href) != 1L || !nzchar(href)) {
      .taco_stop("version '%s' needs a non-empty href", version)
    }
    collection <- .json_object(
      entry[["collection"]],
      sprintf("version '%s' collection", version)
    )
    if (!identical(collection[["dataset_version"]], version)) {
      embedded_version <- collection[["dataset_version"]]
      if (is.null(embedded_version)) embedded_version <- "<missing>"
      .taco_stop(
        "version '%s' embeds collection dataset_version '%s'",
        version,
        as.character(embedded_version)
      )
    }
    entries[[version]] <- list(href = href, collection = collection)
  }

  selected_entry <- entries[[selected]]
  .validate_embedded_collection(selected_entry$collection, selected)
  list(
    sources = .join_manifest_href(candidate, selected_entry$href),
    collection = selected_entry$collection,
    version = selected,
    versions = names(versions),
    manifest = candidate
  )
}
